File: workshops/fullstack-aws/projects/02-notice-board/GervonteFowler/server/repositories/noticeRepository.ts
import { ObjectId } from 'mongodb';
import { getDatabase } from '../db/mongo';
import { Notice, NoticeDocument, toNotice } from '../models/notice';

async function noticesCollection() {
  const database = await getDatabase();
  return database.collection<NoticeDocument>('notices');
}

export async function listNotices(): Promise<Notice[]> {
  const collection = await noticesCollection();
  const documents = await collection.find({}).sort({ created_at: -1 }).toArray();
  return documents.map((document) => toNotice(document as NoticeDocument & { _id: ObjectId }));
}

export async function createNotice(name: string, message: string): Promise<Notice> {
  const collection = await noticesCollection();
  const document: NoticeDocument = {
    name,
    message,
    created_at: new Date().toISOString(),
  };
  const result = await collection.insertOne(document);
  return toNotice({ ...document, _id: result.insertedId });
}

export async function deleteNotice(id: string): Promise<boolean> {
  if (!ObjectId.isValid(id)) {
    return false;
  }
  const collection = await noticesCollection();
  const result = await collection.deleteOne({ _id: new ObjectId(id) });
  return result.deletedCount === 1;
}

File: workshops/fullstack-aws/projects/02-notice-board/pranav-balu-notice-board/backend/app/repositories/notice_repository.py
from bson import ObjectId
from app.database.mongodb import get_database


class NoticeRepository:

    def __init__(self):
        self.collection = get_database()["notices"]

    def get_all(self):
        notices = self.collection.find()

        return [
            {
                "id": str(notice["_id"]),
                "name": notice["name"],
                "message": notice["message"],
            }
            for notice in notices
        ]

    def create(self, name: str, message: str):
        notice = {
            "name": name,
            "message": message,
        }

        result = self.collection.insert_one(notice)

        return {
            "id": str(result.inserted_id),
            "name": name,
            "message": message,
        }

    def delete(self, notice_id: str) -> bool:
        if not ObjectId.is_valid(notice_id):
            return False

        result = self.collection.delete_one(
            {"_id": ObjectId(notice_id)}
        )

        return result.deleted_count > 0